src/addrman.cpp

```diff
@@ -713,37 +713,49 @@ CAddrInfo CAddrMan::Select_(bool newOnly) const
         // use a tried node
         double fChanceFactor = 1.0;
         while (1) {
+            // Pick a tried bucket, and an initial position in that bucket.
             int nKBucket = insecure_rand.randrange(ADDRMAN_TRIED_BUCKET_COUNT);
             int nKBucketPos = insecure_rand.randrange(ADDRMAN_BUCKET_SIZE);
-            while (vvTried[nKBucket][nKBucketPos] == -1) {
-                nKBucket = (nKBucket + insecure_rand.randbits(ADDRMAN_TRIED_BUCKET_COUNT_LOG2)) % ADDRMAN_TRIED_BUCKET_COUNT;
-                nKBucketPos = (nKBucketPos + insecure_rand.randbits(ADDRMAN_BUCKET_SIZE_LOG2)) % ADDRMAN_BUCKET_SIZE;
+            // Iterate over the positions of that bucket.
+            for (int i = 0; i < ADDRMAN_BUCKET_SIZE; ++i) {
+                if (vvTried[nKBucket][nKBucketPos] != -1) {
+                    int nId = vvTried[nKBucket][nKBucketPos];
+                    const auto it_found{mapInfo.find(nId)};
+                    assert(it_found != mapInfo.end());
+                    const CAddrInfo& info{it_found->second};
+                    if (insecure_rand.randbits(30) < fChanceFactor * info.GetChance() * (1 << 30)) return info;
+                    fChanceFactor *= 1.2;
+                    break;
+                }
+                // If the position was empty, transition to the next one (looping around).
+                nKBucketPos = (nKBucketPos + 1) % ADDRMAN_BUCKET_SIZE;
             }
-            int nId = vvTried[nKBucket][nKBucketPos];
-            const auto it_found{mapInfo.find(nId)};
-            assert(it_found != mapInfo.end());
-            const CAddrInfo& info{it_found->second};
-            if (insecure_rand.randbits(30) < fChanceFactor * info.GetChance() * (1 << 30))
-                return info;
-            fChanceFactor *= 1.2;
+            // Either the chosen bucket was empty, or the GetChance()-based bias rejected
```

Comment:
nit: If you move this comment right before the `// Iterate over the positions of that bucket.` comment, next readers won't be like me trying to figure out what exactly this bias tries to achieve.

Reply:
I've changed it back to the original approach, but with iteration instead of xor.